add vertex validation to UndirectedGraphImpl

leblowl opened the pull request:
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
Fixes #1845

leblowl commented on iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/BusBreakerVoltageLevel.java:
```diff
@@ -41,6 +42,13 @@
+    public static final int NODE_LIMIT = PlatformConfig.defaultConfig()
+            .getOptionalModuleConfig("iidm")
+            .map(moduleConfig -> moduleConfig.getIntProperty("node-limit", DEFAULT_NODE_LIMIT))
+            .orElse(DEFAULT_NODE_LIMIT);
```
@flo-dup I tried to copy the example you provided in #1845, but not totally sure if this is correct.

flo-dup replied:
seems good to me!

flo-dup replied:
seems good to me!

leblowl commented on iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/BusBreakerVoltageLevelTest.java:
```diff
@@ -0,0 +1,85 @@
+ * @author Florian Dupuy <florian.dupuy at rte-france.com>
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class BusBreakerVoltageLevelTest {
```
This test is copied from the powsybl-tutorials repo. I couldn't find existing BusBreakerVoltageLevel tests, so I added this file.

leblowl commented on iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/BusBreakerVoltageLevelTest.java:
```diff
@@ -0,0 +1,85 @@
+    @Test
+    public void testBusBreakerVoltageLevelConfig() {
+        VoltageLevel vl = createVoltageLevel();
+        assertEquals(5, BusBreakerVoltageLevel.NODE_LIMIT);
```
@flo-dup I am not really sure how I can test the configuration here. I was thinking it would be good to configure the node-limit and verify that the BusBreakerVoltageLevel is using the correct value.

leblowl commented on iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/NodeBreakerVoltageLevelTest.java:
```diff
@@ -0,0 +1,146 @@
+ * @author Florian Dupuy <florian.dupuy at rte-france.com>
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class NodeBreakerVoltageLevelTest {
```
This test is copied from the powsybl-tutorials repo. I couldn't find existing NodeBreakerVoltageLevel tests, so I added this file.

flo-dup commented on math/src/test/java/com/powsybl/math/graph/UndirectedGraphImplTest.java:
```diff
@@ -76,6 +78,26 @@ public void testAddVertex() {
+        graph.addVertexIfNotPresent(0);
+        assertEquals(1, graph.getVertexCount());
+        graph.addVertexIfNotPresent(0);
+        assertEquals(1, graph.getVertexCount());
```
Maybe also add `VERTEX_LIMIT - 1`

flo-dup reviewed (changes requested):
Good job! Just a small refactor if you agree, and then you could add the config test you suggested

leblowl commented on iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/BusBreakerVoltageLevel.java:
```diff
@@ -41,6 +42,13 @@
 
+    private static final int DEFAULT_NODE_LIMIT = 1000;
+
+    public static final int NODE_LIMIT = PlatformConfig.defaultConfig()
```
Yes, I think that's a better place for it and makes it easier to test. Thanks!

leblowl commented:
@flo-dup I made some changes in response to your feedback, I hope the changes align with what you were thinking.

I moved the node index limit properties and configuration code to the AbstractVoltageLevel class and am setting the NODE_INDEX_LIMIT property statically with a new static `loadNodeIndexLimit` method: https://github.com/powsybl/powsybl-core/blob/9c415ecb4a9a9e1222aeb4d11bef750b614a5956/iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractVoltageLevel.java#L30

I was looking at how some other classes are using the configuration and some are using this static initializer approach: https://github.com/powsybl/powsybl-core/blob/76d5a4b8e209837cb8b4d612b9b5edca2365a0b5/security-analysis/security-analysis-api/src/main/java/com/powsybl/security/LimitViolationFilter.java#L43-L58
We could do that too, it just seemed like it would require a little more refactoring, for example here: https://github.com/powsybl/powsybl-core/blob/76d5a4b8e209837cb8b4d612b9b5edca2365a0b5/iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/VoltageLevelAdderImpl.java#L99-L108

I removed the Node/Bus VoltageLevel tests and replaced them with a single AbstractVoltageLevel test which tests the `loadNodeIndexLimit` method currently.

I made some small wording/test changes as suggested too.

flo-dup reviewed (changes requested):
Great! could you just correct your signoff, so that it passes the DCO test? The first commit is signed off with @mailbox.com but expects @gmail.com, and the second one is not signed off. Then we could have a look at the sonar scan results.

miovd commented on iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractVoltageLevel.java:
```diff
@@ -24,6 +25,10 @@
 
+    private static final int DEFAULT_NODE_INDEX_LIMIT = 1000;
+
+    public static final int NODE_INDEX_LIMIT = loadNodeIndexLimit(PlatformConfig.defaultConfig());
```
Isn't there some risks doing that? It means that the `NODE_INDEX_LIMIT` is fixed when creating the voltage level but what if the user changes it? However, not sure if this is really relevant as it is not a parameter that should change a lot... What do you think @flo-dup @geofjamg ?

flo-dup replied:
What I had in mind was a high number which would fit in most if not all cases - not sure 1000 is a good enough candidate for that. And then it should be hardly ever changed, only for very specific network with huge voltage levels. Or with voltage levels of any size but poorly numbered, that is, with high indices.

miovd replied:
Even in this case, shouldn't then this parameter not be static?

miovd reviewed (commented):
Thank you for your work! I have a question about the definition of `NODE_INDEX_LIMIT` from the configuration but otherwise, great execution.

leblowl commented:
@flo-dup It looks like I got the DCO passing, but I am not seeing the Sonar scan running. Is there something I can do?

flo-dup commented:
> @flo-dup It looks like I got the DCO passing, but I am not seeing the Sonar scan running. Is there something I can do?

I haven't experienced that so far, strange... I'm wondering if it's not because you forked powsybl-core. Maybe if I dig into sonar settings, or otherwise a solution would be for you to create a copycat pull request on a branch of powsybl-core itself? You should have the rights to do it.

leblowl commented:
I created a new branch in powsybl-core and opened a copy PR: https://github.com/powsybl/powsybl-core/pull/1940 which seemed to start the other validation steps, which also got reflected in this PR. Since the checks seem to be tied to the commit, we can just keep using this PR and I can remove the copy or vice versa. For the future, I can create a branch in powsybl-core

flo-dup commented on iidm/iidm-impl/src/test/java/com/powsybl/iidm/network/impl/TestAbstractVoltageLevel.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright (c) 2021, RTE (http://www.rte-france.com)
```
Sorry for unresolving, but in fact after asking a colleague more experienced in that matter, it should rather be your name directly instead of "powsybl authors". Plus it doesn't really make sense to put "Powsybl authors" only in this file and not in the others...

leblowl replied:
No worries, fixed!
